src/server.c

```diff
@@ -3252,10 +3231,65 @@ void InitServerLast() {
     server.initial_memory_usage = zmalloc_used_memory();
 }
 
+/* The purpose of this function is to try to "glue" consecutive range
+ * key specs in order to build the legacy (first,last,step) spec
+ * used by the COMMAND command.
+ * By far the most common case is just one range spec (e.g. SET)
+ * but some commands' ranges were split into two or more ranges
+ * in order to have different flags for different keys (e.g. SMOVE,
+ * first key is "read write", second key is "write").
+ *
+ * This functions uses very basic heuristics and is "best effort":
+ * 1. Only commands which have only "range" specs are considered
+ * 2. Only range specs with keystep of 1 are considered
+ * 3. The order of the range specs must be ascending (i.e.
+ *    lastkey of spec[i] == firstkey-1 of spec[i+1])
+ *
+ * Returns C_OK on success*/
+int populateCommandLegacyRangeSpec(struct redisCommand *c) {
```

Comment:
i'm ok with this function returning `void` (because of what Oran said) but the assertions are necessary IMHO: they make sure the function works properly (i.e. if we have a bug and for some reason lastkey==0 we would like to know about it ASAP rather then fail silently)

for the record: C_OK means that the legacy range was populated correctly (also if we didn't do anything, if there aren't any specs at all)

Reply:
also, it can help a module writer catch bugs just on startup... For example, the module writer provided "index" BS of 0 (which is the command name). we would like to catch that ASAP

Reply:
> i do see that some of the early exists are C_OK (even ones that don't store a legacy spec), and others are C_ERR.

is that a bug then?

Reply:
no, that's how the function was supposed to behave
anyway, it's going to return `void` so it doesn't matter

what do you think about thee assertions? if you're again i could make it return error if the user input doesn't make sense and c_ok otherwise. then the caller needs to fail OnLoad or something like that instead of crashing the server

Reply:
for redis, i suppose an assertion is ok. for a module, i think i'd rather return an error, unless we think it is likely that the module will not check for that error, and the whole thing can go overlooked, in which case i'm ok with an assert (since this is a new API, we'll fail modules during development).

maybe instead of an assertion we can disable module commands that failed that, so the module author will not be able to overlook it, but then it'll be hard for them to figure out why the command got disabled, so i guess an assertion is better.